wasm/src/lsp/patch.rs

```diff
@@ -0,0 +1,192 @@
+use std::borrow::Borrow;
+use std::collections::HashMap;
+use std::rc::Rc;
+
+use lsp_types::TextDocumentContentChangeEvent;
+
+use crate::lsp::parse::DocVecByteIter;
+use crate::lsp::types::DocData;
+use crate::lsp::LSPServiceProvider;
+
+pub trait PatchableDocument {
+    fn apply_patch(&self, version: i32, patches: &[TextDocumentContentChangeEvent]) -> Self;
+}
+
+pub trait LSPServiceProviderApplyDocumentPatch {
+    fn apply_document_patch(
+        &mut self,
+        uristring: &str,
+        version: i32,
+        patches: &[TextDocumentContentChangeEvent],
+    );
+}
+
+pub fn split_text(td: &str) -> Vec<Rc<Vec<u8>>> {
```

Comment:
I think it would be good to have unit tests for these functions (and really, all functions, whether they have external or internal linkage). I don't see any tests in this file, but it's a bit hard to get a full overview of this patch on github.

This comment applies to a lot of other files too, that have functions without unit tests (that I can find at least)

Reply:
Ya i will get in the habit of moving tests into the files they go along with.  The rust structure i've been using puts tests in a separate directory to make it clear that they and their associated infrastructure don't contribute to binary size).  I find that reasonable but from a review perspective I'm sure it's different.

Lots of this is tested for the most part in integration tests that show the outcome of various things, but as I have said I know it isn't fully complete yet.